src/Avalonia.Controls/ComboBox.cs

```diff
@@ -0,0 +1,192 @@
+        }
+
+        /// <inheritdoc/>
+        protected override IItemContainerGenerator CreateItemContainerGenerator()
+        {
+            return new ItemContainerGenerator<ComboBoxItem>(
+                this,
+                ComboBoxItem.ContentProperty,
+                ComboBoxItem.ContentTemplateProperty);
+        }
+
+        /// <inheritdoc/>
+        protected override void OnAttachedToLogicalTree(LogicalTreeAttachmentEventArgs e)
+        {
+            base.OnAttachedToLogicalTree(e);
+            this.UpdateSelectionBoxText(this.SelectedItem);
+        }
+
+        /// <inheritdoc/>
+        protected override void OnKeyDown(KeyEventArgs e)
+        {
+            base.OnKeyDown(e);
+
+            if (!e.Handled)
+            {
+                if (e.Key == Key.F4 ||
+                    (e.Key == Key.Down && ((e.Modifiers & InputModifiers.Alt) != 0)))
+                {
+                    IsDropDownOpen = !IsDropDownOpen;
+                    e.Handled = true;
+                }
+                else if (IsDropDownOpen && (e.Key == Key.Escape || e.Key == Key.Enter))
+                {
+                    IsDropDownOpen = false;
+                    e.Handled = true;
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void OnPointerPressed(PointerPressedEventArgs e)
+        {
+            if (!IsDropDownOpen && ((IVisual)e.Source).GetVisualRoot() != typeof(PopupRoot))
+            {
+                IsDropDownOpen = true;
+                e.Handled = true;
+            }
+
+            if (!e.Handled)
+            {
+                if (UpdateSelectionFromEventSource(e.Source))
+                {
+                    _popup?.Close();
+                    e.Handled = true;
+                }
+            }
+
+            base.OnPointerPressed(e);
+        }
+
+        /// <inheritdoc/>
+        protected override void OnTemplateApplied(TemplateAppliedEventArgs e)
+        {
+            if (_popup != null)
+            {
+                _popup.Opened -= PopupOpened;
+            }
+
+            _popup = e.NameScope.Get<Popup>("PART_Popup");
+            _popup.Opened += PopupOpened;
+
+            if (_textBox != null)
+            {
+                _textBox.TextInput -= TextBoxInput;
+            }
+
+            _textBox = e.NameScope.Get<TextBox>("PART_TextBox");
+            _textBox.TextInput += TextBoxInput;
+        }
+
+        private void PopupOpened(object sender, EventArgs e)
+        {
+            var selectedIndex = SelectedIndex;
+
+            if (selectedIndex != -1)
+            {
+                var container = ItemContainerGenerator.ContainerFromIndex(selectedIndex);
+                container?.Focus();
+            }
+        }
+
+        private void TextBoxInput(object sender, TextInputEventArgs e)
+        {
+            int i = 0;
+
+            foreach (var item in Items)
+            {
+                if (item.ToString() == _textBox.Text)
```

Comment:
If there's no exact match, the first item that starts with the text should be selected. For example, if the list contains "Item1", "Item2", "Item33", , "Item34", when typing "I", "Item1" would be selected, and when typing "Item3", "Item 33" would be selected. If possible, please unit test those.